### ISSUE TITLE
Building Glibc with -Og Fails

### ISSUE THREAD

**Issue by shahab-vahedi** [foss-for-synopsys-dwc-arc-processors/gcc#79]
Tools versions:
```
binutils:  arc-2019.09
gcc:       arc-2018.09
glibc:     arc-glibc-master
toolchain: arc-dev

gdb:       arc-2019.09-gdb
linux:     qemu-5.1
newlib:    arc-2019.03
uclibc-ng: master
```

Build command:
```
build-all.sh --source-dir    /src/arcgnu     \
             --build-dir     /build/arcgnu/  \
             --install-dir   /install/arcgnu \
             --release-name  "201909 glibc"  \
             --cpu           archs           \
             --no-multilib                   \
             --no-uclibc                     \
             --glibc                         \
             --target-cflags "-Og -g3"       \
             --jobs          8               \
             --no-pdf                        \
             --no-auto-checkout              \
             --no-auto-pull                  \
             --no-external-download          \
             --no-strip                      \
             --native-gdb                    \
             --system-expat                  \
             --no-elf32-strip-target-libs    \
             --no-native                     \
             --no-optsize-newlib             \
             --no-optsize-libstdc++
```

Failure output:
```
In file included from regex.c:72:
regexec.c: In function ‘check_node_accept_bytes’:
regexec.c:3843:29: warning: ‘extra’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 3843 |        const unsigned char *coll_sym = extra + cset->coll_syms[i];
      |                             ^~~~~~~~
/tmp/cc5yIPGY.s: Assembler messages:
/tmp/cc5yIPGY.s:5522: Error: Insn b has an instruction mov_s with limm in its delay slot.
/tmp/cc5yIPGY.s:13921: Error: Insn b has an instruction mov_s with limm in its delay slot.
```

**compiler command used for compiling `regex.c`:**
```
arc-snps-linux-gnu-gcc regex.c                                                                        \
                       -c                                                                             \
                       -std=gnu11                                                                     \
                       -fgnu89-inline                                                                 \
                       -Og                                                                            \
                       -g3                                                                            \
                       -Wall                                                                          \
                       -Wwrite-strings                                                                \
                       -Wundef                                                                        \
                       -fmerge-all-constants                                                          \
                       -frounding-math                                                                \
                       -fno-stack-protector                                                           \
                       -Wstrict-prototypes                                                            \
                       -Wold-style-definition                                                         \
                       -fmath-errno                                                                   \
                       -ftls-model=initial-exec                                                       \
                       -I../include                                                                   \
                       -I/build/arcgnu/bd-glibc/glibc/posix                                           \
                       -I/build/arcgnu/bd-glibc/glibc                                                 \
                       -I../sysdeps/unix/sysv/linux/arc                                               \
                       -I../sysdeps/arc/nptl                                                          \
                       -I../sysdeps/unix/sysv/linux/generic/wordsize-32                               \
                       -I../sysdeps/unix/sysv/linux/generic                                           \
                       -I../sysdeps/unix/sysv/linux/include                                           \
                       -I../sysdeps/unix/sysv/linux                                                   \
                       -I../sysdeps/nptl                                                              \
                       -I../sysdeps/pthread                                                           \
                       -I../sysdeps/gnu                                                               \
                       -I../sysdeps/unix/inet                                                         \
                       -I../sysdeps/unix/sysv                                                         \
                       -I../sysdeps/unix                                                              \
                       -I../sysdeps/posix                                                             \
                       -I../sysdeps/arc/nofpu                                                         \
                       -I../sysdeps/ieee754/soft-fp                                                   \
                       -I../sysdeps/arc                                                               \
                       -I../sysdeps/init_array                                                        \
                       -I../sysdeps/wordsize-32                                                       \
                       -I../sysdeps/ieee754/flt-32                                                    \
                       -I../sysdeps/ieee754/dbl-64                                                    \
                       -I../sysdeps/ieee754                                                           \
                       -I../sysdeps/generic                                                           \
                       -I..                                                                           \
                       -I../libio                                                                     \
                       -I.                                                                            \
                       -nostdinc                                                                      \
                       -isystem /install/arcgnu/bin/../lib/gcc/arc-snps-linux-gnu/9.2.1/include       \
                       -isystem /install/arcgnu/bin/../lib/gcc/arc-snps-linux-gnu/9.2.1/include-fixed \
                       -isystem /install/arcgnu/arc-snps-linux-gnu/sysroot/usr/include                \
                       -D_LIBC_REENTRANT                                                              \
                       -include /build/arcgnu/bd-glibc/glibc/libc-modules.h                           \
                       -DMODULE_NAME=libc                                                             \
                       -include ../include/libc-symbols.h                                             \
                       -DTOP_NAMESPACE=glibc                                                          \
                       -o /build/arcgnu/bd-glibc/glibc/posix/regex.o                                  \
                       -MD -MP -MF /build/arcgnu/bd-glibc/glibc/posix/regex.o.dt                      \
                       -MT /build/arcgnu/bd-glibc/glibc/posix/regex.o
```

**If you replace `-Og` with `-O2`, it will compile OK.**

**Comment by shahab-vahedi** [comment]
problematic assembly (1):
```
.LVL539:
	.loc 5 3167 11 view .LVU1733
	extb.f	r0,r0
	bne.d @.L422
	mov_s.ne r0,12
	b_s @.L422
	.align 4
```

problematic assembly (2):
```
.LVL1396:
	.loc 4 1531 11 view .LVU4231
	extb.f	r0,r0
	bne.d @.L1021
	mov_s.ne r0,12
	b_s @.L1021
.LVL1397:
	.align 4
```

**Comment by shahab-vahedi** [comment]
[reduced](https://embed.cs.utah.edu/creduce/) C code:
```
typedef struct
{
  struct
  {
    int length
  } table
} room;

struct house
{
  room *r
};

undefined;

_Bool bar();

func(struct house *h, int i, int whatever)
{
  for (;;)
  {
    _Bool a;
    if (h && h->r[i].table.length == undefined)
    {
      if (whatever)
      {
        a = bar();
        if (__builtin_expect(!a, 0))
          return 7;
      }
      break;
    }
  }
}
```

compile with:
```
cc1 test.c -mcpu=archs -Og -o test.s
as test.s -o test.o
```

output:
```
test.s: Assembler messages:
test.s:37: Error: Insn b has an instruction mov_s with limm in its delay slot.
```

problem:
```
bne.d @.L1      # 33    [c=20 l=4]  *branch_insn
mov_s.ne r0,7   # 35    [c=0 l=6]  *movsi_ne/2
```

**Comment by shahab-vahedi** [comment]
`-fno-delayed-branch` by-passes this problem.

**Comment by claziss** [comment]
The issue is due to a mov instruction which is placed into a annul-true delay slot. Hence, it needs to be predicated. As it is having an immediate, it ends up in selecting `mov_s.ne` instruction of the `movsi_ne` pattern. However, this pattern choices list is ordered against what internal attribute `cond` assumes, hence, leading to the selection of the `mov_s.ne` mnemonic instead of the `mov.ne` mnemonic. Fix in 53803396251d58fd664b9fba8c30399972908756.

**Comment by shahab-vahedi** [comment]
I confirm: it is fixed like a charm :)
